core/cog-viewport.c

```diff
@@ -60,6 +65,24 @@ G_DEFINE_TYPE_WITH_PRIVATE(CogViewport, cog_viewport, G_TYPE_OBJECT)
 
 #define PRIV(obj) ((CogViewportPrivate *) cog_viewport_get_instance_private((CogViewport *) obj))
 
+static void *
+_cog_viewport_get_impl_type_init(GType *type)
+{
+    *type = COG_TYPE_VIEWPORT;
+
+    CogPlatform *platform = cog_platform_get();
+    if (platform) {
+        CogPlatformClass *platform_class = COG_PLATFORM_GET_CLASS(platform);
+        if (platform_class->get_viewport_type) {
+            GType viewport_type = (*platform_class->get_viewport_type)();
+            *type = viewport_type;
```

Comment:
When doing the follow-up patch with cleanups, I think we could assert here that we get a valid type:

```suggestion
            GType viewport_type = (*platform_class->get_viewport_type)();
            g_assert(viewport_type != G_TYPE_INVALID);
            *type = viewport_type;
```

Reply:
This looks good I will add a note in  https://github.com/Igalia/cog/issues/664 to address this here in the CogViewport but also in the CogView